packages/cojson/src/streamUtils.ts

```diff
@@ -1,47 +1,89 @@
-import { Channel } from "queueueue";
-import { Peer, PeerID, SyncMessage } from "./sync.js";
-export { Channel } from "queueueue";
+import {
+  DisconnectedError,
+  IncomingPeerChannel,
+  OutgoingPeerChannel,
+  Peer,
+  PeerID,
+  SyncMessage,
+} from "./sync.js";
 
 export function connectedPeers(
   peer1id: PeerID,
   peer2id: PeerID,
   {
     peer1role = "client",
     peer2role = "client",
-    crashOnClose = false,
   }: {
     peer1role?: Peer["role"];
     peer2role?: Peer["role"];
-    crashOnClose?: boolean;
   } = {},
 ): [Peer, Peer] {
-  const [from1to2Rx, from1to2Tx] = newQueuePair();
-  const [from2to1Rx, from2to1Tx] = newQueuePair();
+  const from1to2 = new ConnectedPeerChannel();
+  const from2to1 = new ConnectedPeerChannel();
 
   const peer2AsPeer: Peer = {
     id: peer2id,
-    incoming: from2to1Rx,
-    outgoing: from1to2Tx,
+    incoming: from2to1,
+    outgoing: from1to2,
     role: peer2role,
-    crashOnClose: crashOnClose,
   };
 
   const peer1AsPeer: Peer = {
     id: peer1id,
-    incoming: from1to2Rx,
-    outgoing: from2to1Tx,
+    incoming: from1to2,
+    outgoing: from2to1,
     role: peer1role,
-    crashOnClose: crashOnClose,
   };
 
   return [peer1AsPeer, peer2AsPeer];
 }
 
-export function newQueuePair(): [
-  AsyncIterable<SyncMessage>,
-  Channel<SyncMessage>,
-] {
-  const channel = new Channel<SyncMessage>();
+export function newQueuePair(): [ConnectedPeerChannel, ConnectedPeerChannel] {
+  const channel = new ConnectedPeerChannel();
 
-  return [channel.wrap(), channel];
+  return [channel, channel];
+}
+
+export class ConnectedPeerChannel
+  implements IncomingPeerChannel, OutgoingPeerChannel
+{
+  buffer: (SyncMessage | DisconnectedError)[] = [];
```

Comment:
Changed the Peer channels to use a callback-based communication instead of relying on async iterators.

Used a buffer to collect all the messages pushed before the connection with the syncManager.